feat: increment an internal metric when duplicate validation errors are found

gavinkflam opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it:

Several resources currently emit duplicate validation errors during update operations. This results in bloated and confusing API responses and redundant processing. I've submitted a few PRs to address specific cases, but there are likely many more instances still present in the codebase.

Detecting these duplicates is challenging because kubectl deduplicates error messages before displaying them to the user and hiding the issue.

This PR introduces an internal metric and a log message to track the duplicate validation errors. It provides visibility into remaining duplicates and help to identify opportunities for further cleanup.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

Fixes #130656

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

gavinkflam commented:
/remove-area stable-metrics

gavinkflam commented:
Changed the metric to alpha according to [KEP-1209](https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1209-metrics-stability).

thockin commented on staging/src/k8s.io/apiserver/pkg/registry/rest/validate.go:
```diff
@@ -320,3 +321,23 @@ func panicSafeValidateFunc(
+// and increment duplicate validation error metric when duplicates are found.
+func DeduplicateValidationErrorsAndUpdateMetric(qualifiedKind schema.GroupKind, operation string, errs field.ErrorList) field.ErrorList {
+	uniqueErrs := field.ErrorList{}
+	errsSet := sets.New[string]()
```
given how few errors there are, we should just `make([]string, 0, len(errs))`, do `slices.Contains()` for lookup, and append to the list if it is not found.

gavinkflam replied:
thanks. this is updated.

thockin commented on staging/src/k8s.io/apiserver/pkg/validation/metrics.go:
```diff
@@ -52,6 +53,15 @@ var validationMetricsInstance = &validationMetrics{
 			StabilityLevel: metrics.BETA,
 		},
 	),
+	DuplicateValidationErrorCounter: metrics.NewCounterVec(
```
@jpbetz -- does this need a KEP?

jpbetz replied:
I'm not 100% sure.  @richabanker do you know?  Can we introduce a metric without a KEP?

richabanker replied:
yeah metrics are allowed to be added without a KEP, have seen multiple cases like that in the past.

richabanker replied:
but.. this metric's purpose seems to be to identify potential areas for further code-cleanup? Not sure if thats super useful to the end user? Can have this as an [INTERNAL](https://github.com/kubernetes/kubernetes/blob/release-1.33/staging/src/k8s.io/component-base/metrics/opts.go#L85) type metric maybe.

gavinkflam replied:
Good idea. I can make this internal.

richabanker commented:
/ok-to-test

thockin commented on staging/src/k8s.io/apiserver/pkg/registry/rest/validate.go:
```diff
@@ -320,3 +321,23 @@ func panicSafeValidateFunc(
+
+		if slices.Contains(seenErrs, errStr) {
+			// Increment the duplicate error metric counter
+			validationmetrics.Metrics.IncDuplicateValidationErrorMetric(qualifiedKind.String(), operation)
```
I am not a prometheus expert but isn't this potentially A LOT of cardinality?

Why not log the details and let the metric indicate "it happened, go look at the logs" ?

gavinkflam replied:
That makes sense. Logs are more helpful here too. We can log the entire error, which includes the field path and other useful context.

gavinkflam replied:
Just pushed an update - thanks for the advise!

thockin reviewed (commented):
Given that kubectl DOES do this filtering, do we need to remove the dup errors here, or just flag that dups happened?  It saves one allocation at least.

Is there ever a future where this `Dedup...` code gets removed?  I am guessing not, so making it as low-impact as possible seems ideal.  In 99.999% of cases the original list is unique.

gavinkflam commented on staging/src/k8s.io/apiserver/pkg/registry/rest/validate.go:
```diff
@@ -320,3 +321,25 @@ func panicSafeValidateFunc(
+	seenErrs := make([]string, 0, len(errs))
+
+	for _, err := range errs {
+		errStr := fmt.Sprintf("%+v", err)
```
You are right. It's the same with or without `+`. I've removed it.

gavinkflam commented on staging/src/k8s.io/apiserver/pkg/registry/rest/validate_test.go:
```diff
@@ -678,6 +681,74 @@ func TestValidateUpdateDeclarativelyWithRecovery(t *testing.T) {
+			operation:     "UPDATE",
+			qualifiedKind: schema.GroupKind{Group: "apps", Kind: "ReplicaSet"},
+			errs: field.ErrorList{
+				field.Invalid(field.NewPath("spec").Child("replicas"), -1, "must be greater than or equal to 0").WithOrigin("minimum"),
```
Added a test case for this.

gavinkflam committed:
feat: increment an internal metric when duplicate validation errors

Signed-off-by: Gavin Lam <gavin.oss@tutamail.com>

gavinkflam commented:
@thockin I agree it's best not to dedup here. I've addressed your feedbacks and simplified the code.

Please take another look if you get a chance. Thank you for the thoughtful review and feedbacks!

richabanker commented on staging/src/k8s.io/apiserver/pkg/registry/rest/validate.go:
```diff
@@ -320,3 +321,20 @@ func panicSafeValidateFunc(
+// RecordDuplicateValidationErrors increments a metric and log the error when duplicate validation errors are found.
+func RecordDuplicateValidationErrors(ctx context.Context, qualifiedKind schema.GroupKind, errs field.ErrorList) {
+	logger := klog.FromContext(ctx)
+	seenErrs := make([]string, 0, len(errs))
```
is it worth having this as a map instead?

gavinkflam replied:
@thockin advised to use slices given the small amount of errors here.

https://github.com/kubernetes/kubernetes/pull/132613#discussion_r2176573089

thockin commented on staging/src/k8s.io/apiserver/pkg/registry/rest/validate.go:
```diff
@@ -320,3 +321,25 @@ func panicSafeValidateFunc(
+		if slices.Contains(seenErrs, errStr) {
+			logger.Info("Removed duplicate validation error", "kind", qualifiedKind.String(), "operation", operation, "error", errStr)
+			// Increment the duplicate error metric counter
+			validationmetrics.Metrics.IncDuplicateValidationErrorMetric()
```
also @BenTheElder

BenTheElder replied:
ACK -- I want to ship something very lightweight for declarative validation first since that is blocking the testing plan for that KEP and iterate from there, x-ref https://github.com/kubernetes/enhancements/issues/5196

thockin replied:
That's true, but the transition might be a long time.

BenTheElder replied:
Will circle back soon with a KEP https://github.com/kubernetes/enhancements/issues/5468

thockin reviewed (commented):
Thanks!

/lgtm
/approve
